File: 04/problem_04.py
import sys


n_contains = 0
n_overlaps = 0

with open(sys.argv[1], "rt") as input_file:
    for line in input_file:
        r1, _, r2 = line.strip().partition(',')
        r1 = [int(v) for v in r1.partition('-')[::2]]
        r2 = [int(v) for v in r2.partition('-')[::2]]
        if r2[0] >= r1[0] and r2[1] <= r1[1]:
            # r1 contains r2
            n_contains += 1
        elif r1[0] >= r2[0] and r1[1] <= r2[1]:
            # r2 contains r1
            n_contains += 1
        if r2[0] <= r1[1] and r2[1] >= r1[0]:
            n_overlaps += 1
        elif r1[0] <= r2[1] and r1[1] >= r2[0]:
            n_overlaps += 1

print(n_contains)
print(n_overlaps)

File: 04/problem_04.rs
use std::io;
use std::io::prelude::*;
use std::io::BufReader;
use std::fs::File;


fn main() -> io::Result<()> {
    let f: File = File::open("input.txt")?;
    let reader: BufReader<File> = BufReader::new(f);

    let mut n_contains: i32 = 0;
    let mut n_overlaps: i32 = 0;

    for line_ in reader.lines() {
        let line: String = line_?;
        let (r1, r2) = line.split_once(",").unwrap();
        let r1r: Vec<i32> = r1.split("-").map(|v| v.parse::<i32>().unwrap()).collect();
        let r2r: Vec<i32> = r2.split("-").map(|v| v.parse::<i32>().unwrap()).collect();
        if (r2r[0] >= r1r[0]) && (r2r[1] <= r1r[1]) {
            n_contains += 1;
        } else if (r1r[0] >= r2r[0]) && (r1r[1] <= r2r[1]) {
            n_contains += 1;
        }
        if (r2r[1] >= r1r[0]) && (r2r[0] <= r1r[1]) {
            n_overlaps += 1;
        } else if (r1r[1] >= r2r[0]) && (r1r[0] <= r2r[1]) {
            n_overlaps += 1;
        }
    }

    println!("{}", n_contains);
    println!("{}", n_overlaps);

    return Ok(());
}